### DISCUSSION TITLE
how to handle an auto-complete interaction error?

### DISCUSSION THREAD

**Discussion by modernNeo** [Rapptz/discord.py#9711] [Q&A]
I noticed today in the logs of my bot that there was this

```
2024-01-19 13:18:53 = ERROR = discord.app_commands.tree = Ignoring exception in autocomplete for 'iam'
Traceback (most recent call last):
.  File "/usr/local/lib/python3.8/site-packages/discord/app_commands/tree.py", line 1240, in _call
.    await command._invoke_autocomplete(interaction, focused, namespace)
.  File "/usr/local/lib/python3.8/site-packages/discord/app_commands/commands.py", line 897, in _invoke_autocomplete
.    await interaction.response.autocomplete(choices)
.  File "/usr/local/lib/python3.8/site-packages/discord/interactions.py", line 1027, in autocomplete
.    await adapter.create_interaction_response(
.  File "/usr/local/lib/python3.8/site-packages/discord/webhook/async_.py", line 219, in request
.    raise NotFound(response, data)
discord.errors.NotFound: 404 Not Found (error code: 10062): Unknown interaction
```

and I was wondering if it is possible to handle this error more gracefully?

my bot already has overriden https://discordpy.readthedocs.io/en/stable/api.html#discord.Client.on_error to be something more appropriate for my logging and bot setup but I don't see a way to handle errors that occur when creating a response to an autocomplete callback. Does such a feature not exist?

**Comment by No767** [comment]
You'd have to override the `on_error` coroutine found in `app_commands.CommandTree` (or use the `@error` decorator). `discord.Client.on_error` doesn't catch app commands errors

**Comment by felipeyx** [comment]
Client.on_error does not handle application-command errors because they belong to the CommandTree.

You can provide a custom tree/error handler:

class MyTree(app_commands.CommandTree):
    async def on_error(
        self,
        interaction: discord.Interaction,
        error: app_commands.AppCommandError,
    ):
        print(f"app command error: {error!r}")

However, for autocomplete specifically, 10062 Unknown interaction usually means the autocomplete interaction expired before Discord received the response.

Autocomplete handlers should therefore be extremely fast. Avoid database/API requests inside them when possible. Cache the possible values first, then only filter the cache in the callback:

async def autocomplete(interaction, current: str):
    return [
        app_commands.Choice(name=x, value=x)
        for x in cached_values
        if current.lower() in x.lower()
    ][:25]

So I would handle/log app-command errors through the tree, but treat reducing autocomplete latency as the actual fix for the 10062.
